Performance fixes for larger files

Doublonmousse opened the pull request:
Fixes #1436 
(and maybe #1438, waiting on a response)

- use `push_repeat` for background nodes : This seems to reduce RAM usage and buffer uploads to the gpu so there's more headroom for the rest of the nodes
- use the rtree to speed up `regenerate_rendering_in_viewport_threaded` : We use the rtree to get the strokes in the viewport without recalculating any stroke bounds, and get the ones not in the view by iterating on keys and using a hashmap to filter out the previously selected ones

On that subject, it seems we could use a https://docs.rs/slotmap/1.0.7/slotmap/sparse_secondary/struct.SparseSecondaryMap.html for the render components and/or be smarter on clearing the `render_comp` components of strokes (we know there's no work to do if we have x strokes in view that we just initialized the rendering components and x strokes in the rendering component slot)

I've updated the rtree to only keep bounds for visible strokes (so trashed strokes are removed from the tree)

- [use rtree for update_content_rendering_current_viewport calculations](https://github.com/flxzt/rnote/commit/e75519864ef25c03b0560c06f6044736ccae817e) Once again, using the rtree to calculate document bounds is much faster

I've also added to the visual debug mode a way to see the envelope of the whole rtree (marked by orange bounds)

Doublonmousse committed:
perf: use `push_repeat` for background nodes

This diminishes the cost for this
- one element of somewhat fixed size (no vec pushing)
- one element given to gtk that can tile this as it pleases with less buffer pushes

Doublonmousse committed:
perf: use `push_repeat` for background nodes

This diminishes the cost for this
- one element of somewhat fixed size (no vec pushing)
- one element given to gtk that can tile this as it pleases with less buffer pushes

Doublonmousse committed:
visual debug : show rtree bounds

Doublonmousse committed:
use rtree to speed up `regenerate_rendering_for_strokes_threaded`

We use the rtree to find the element in the viewport, mark these ones using a hashmap, call the rendering function, then remove the rest by iterating on the keys a second time, filtering on keys not in the hashmap

Doublonmousse committed:
use rtree to speed up `regenerate_rendering_for_strokes_threaded`

We use the rtree to find the element in the viewport, mark these ones using a hashmap, call the rendering function, then remove the rest by iterating on the keys a second time, filtering on keys not in the hashmap

Doublonmousse committed:
make the rtree only keep track of non-trashed strokes

Doublonmousse committed:
make the rtree only keep track of non-trashed strokes

Doublonmousse committed:
fix typo

Doublonmousse committed:
fix typo

Doublonmousse committed:
use rtree for `update_content_rendering_current_viewport` calculations

Doublonmousse committed:
use rtree for `update_content_rendering_current_viewport` calculations

Doublonmousse committed:
rebuilt rtree for non trashed components only

Doublonmousse committed:
rebuilt rtree for non trashed components only

Doublonmousse committed:
add a note on rendernode retention when trashing strokes

Doublonmousse committed:
add a note on rendernode retention when trashing strokes

Doublonmousse committed:
clean up interfaces

Do not leak key trees out of the store structure

Doublonmousse committed:
clean up interfaces

Do not leak key trees out of the store structure

Doublonmousse committed:
fix typo

Doublonmousse committed:
fix typo

Doublonmousse committed:
handle correctly empty keytree for `resize_doc` methods

Doublonmousse committed:
handle correctly empty keytree for `resize_doc` methods

flxzt commented on crates/rnote-engine/src/store/keytree.rs:
```diff
@@ -69,6 +70,18 @@ impl KeyTree {
         *self = Self::default()
     }
+
+    pub(crate) fn get_tree(&self) -> &rstar::RTree<KeyTreeObject, rstar::DefaultParams> {
```
The idea was to encapsulate all needed functionality from the R-Tree. Better would be to add the methods operating on the R-Tree itself here instead of exposing it to the rest of the crate

flxzt commented on crates/rnote-engine/src/store/keytree.rs:
```diff
@@ -69,6 +70,18 @@ impl KeyTree {
+        &self.0
+    }
+
+    pub fn get_bounds(&self) -> AABB<[f64; 2]> {
```
RTree's AABB type should not be exposed outside of the module and instead be converted to parry's Aaab

flxzt commented on crates/rnote-engine/src/store/render_comp.rs:
```diff
@@ -6,9 +6,12 @@ use crate::strokes::content::GeneratedContentImages;
 use tracing::error;
 
+#[cfg(feature = "ui")]
+use rnote_compose::shapes::Shapeable;
```
Import dependencies in the guards themselves like it is done in other parts of the code (e.g. L606)

flxzt commented on crates/rnote-engine/src/store/render_comp.rs:
```diff
@@ -242,28 +245,39 @@ impl StrokeStore {
+        let keys_in_viewport = self
+            .key_tree
+            .get_tree()
+            .locate_in_envelope_intersecting(&rstar::AABB::from_corners(
```
use method `keys_intersecting_bounds` defined in keytree:L38 instead of operating on the raw R-Tree here.

Doublonmousse replied:
I can't do that directly as the method has already collected into a vec (and doing an into_iter didn't really work because of type signatures).
The next best thing is a `keys_intersecting_bounds_hash` variant that collects into a HashMap

flxzt commented on crates/rnote-engine/src/store/render_comp.rs:
```diff
@@ -242,28 +245,39 @@ impl StrokeStore {
+        // rtree but also get from this the keys that are not in here
+        // for that also create a slotmap of keys that are in the viewport
+        // so that we can iterate a second time on keys and filter on elements not in the slotmap
+        let mut hashmap_in_viewport: HashMap<StrokeKey, ()> = HashMap::new();
```
instead of collecting the keys as a side-effect of the `.map()` adapter, consider collecting into a `HashSet`. This way we can still only iterate once over all keys of `render_components` and check whether the key is in this hashset before calculating the bounds.

flxzt commented on crates/rnote-engine/src/store/render_comp.rs:
```diff
@@ -656,6 +686,18 @@ impl StrokeStore {
+                na::point![tree_bounds.lower()[0], tree_bounds.lower()[1]],
+                na::point![tree_bounds.upper()[0], tree_bounds.upper()[1]],
+            ),
+            rnote_compose::Color::new(1.0, 0.5, 0., 1.0),
```
don't hardcode the color, instead define a color on module level.

flxzt commented on crates/rnote-engine/src/store/mod.rs:
```diff
@@ -10,6 +10,7 @@ pub mod trash_comp;
 pub use chrono_comp::ChronoComponent;
 use keytree::KeyTree;
 pub use render_comp::RenderComponent;
+use rstar::AABB;
```
move this under `// Imports`

flxzt commented on crates/rnote-engine/src/engine/rendering.rs:
```diff
@@ -35,21 +34,20 @@ impl Engine {
-                for split_bounds in viewport.split_extended_origin_aligned(
+                let origin_aabb = viewport.get_origin(
                     self.document.config.background.tile_size(),
                     SplitOrder::default(),
```
What's the split order used for here? I don't understand why it is an argument in the first place.

flxzt commented on crates/rnote-engine/src/engine/rendering.rs:
```diff
@@ -251,17 +250,23 @@ impl Engine {
+        ) {
+            snapshot.push_repeat(
+                &graphene::Rect::from_p2d_aabb(doc_bounds),
+                Some(&graphene::Rect::from_p2d_aabb(bounds)),
```
it seems to me that we can get by by passing `None` for the child bounds argument here, because then the entire background tile rendernode size is used

flxzt commented on crates/rnote-engine/src/engine/mod.rs:
```diff
@@ -199,7 +199,8 @@ pub struct Engine {
     #[serde(skip)]
-    background_rendernodes: Vec<gtk4::gsk::RenderNode>,
+    background_rendernode: Option<gtk4::gsk::RenderNode>,
+    origin_background_rendernode: Option<Aabb>,
```
*If* this is still needed, this should be renamed and guarded by the ui feature.

flxzt reviewed (commented):
when this lands these are nice improvements to the rendering part of the engine. I added some comments

flxzt commented:
Only storing non-trashed strokes in the R-Tree is problematic and will require additional changes in other parts - essentially everywhere where one of the methods of `KeyTree` are used (like `keys_in_bounds()`, `keys_intersecting_bounds()` etc. Right now they are all used under the assumption that **all** strokes are included in the key. Did you consider this?

Doublonmousse committed:
refactor: move color for rtree bounds to the `visual_debug` module

Doublonmousse commented on crates/rnote-engine/src/store/stroke_comp.rs:
```diff
@@ -127,17 +127,12 @@ impl StrokeStore {
+            return 0.0;
+        } else {
+            let bounds = self.key_tree.get_bounds();
+            bounds.upper()[1] - bounds.lower()[1]
```
Having the keytree only hold the non trashed keys makes it possible to get the bounds in constant time here.

I'm currently not doing that  (only holding non trashed keys in the rtree) so some behavior is modified but I'm unclear of the consequences as of now. The only part I can find is that the vertical spacer tool will not move thrashed keys anymore.

So I'm not sure what's the best way forward.

Doublonmousse replied:
The two keytrees are implemented in https://github.com/flxzt/rnote/pull/1462/commits/5aee84d644f114053d01b8ad27e9ed99967990ff

Doublonmousse committed:
refactor: remove `origin_background_rendernode` (using `None` for the `push_repeat` instead)

`push_repeat(x,None)` will take the whole not and repeat it

Doublonmousse committed:
refactor: remove `origin_background_rendernode` (using `None` for the `push_repeat` instead)

`push_repeat(x,None)` will take the whole not and repeat it

Doublonmousse committed:
refactor: rename `get_origin` method and remove unneeded arguments

Doublonmousse committed:
refactor: rename `get_origin` method and remove unneeded arguments

Doublonmousse committed:
add `keys_intersecting_bounds_hashset` for `render_comp`

Doublonmousse committed:
add `keys_intersecting_bounds_hashset` for `render_comp`

Doublonmousse committed:
refactor: remove `get_tree` method

Doublonmousse committed:
refactor: remove `get_tree` method

Doublonmousse committed:
refactor: do not leak `AABB` outside, use `Aabb` instead

Doublonmousse committed:
refactor: do not leak `AABB` outside, use `Aabb` instead

Doublonmousse committed:
import `rnote_compose::shapes::Shapeable` inside ui guarded functions

Doublonmousse committed:
import `rnote_compose::shapes::Shapeable` inside ui guarded functions

Doublonmousse committed:
refactor: remove extraneous conversion from `Aabb` to itself

Doublonmousse committed:
refactor: remove extraneous conversion from `Aabb` to itself

Doublonmousse committed:
remove code from another feature

that would push a 'current' stroke key to the rendering keys

Doublonmousse committed:
duplicate the rtree for trashed and non trashed keys

And remove now superfluous filters on the trash status for the outputs of the `key_tree`

Doublonmousse committed:
duplicate the rtree for trashed and non trashed keys

And remove now superfluous filters on the trash status for the outputs of the `key_tree`

Doublonmousse committed:
visual debug: show the `trashed_keytree`

Doublonmousse committed:
`trash_colliding_strokes`: get the stroke bounds from the keytree directly

The information is there already, and it is probably more effective to get the value from the keytree rather than recalculating it on the fly on each event

Doublonmousse committed:
`trash_colliding_strokes`: get the stroke bounds from the keytree directly

The information is there already, and it is probably more effective to get the value from the keytree rather than recalculating it on the fly on each event

Doublonmousse committed:
move the tree/rtree calls before the stroke-specific calls

This seems to fix weird graphical glitches that would occur otherwise

Doublonmousse committed:
fmt

Doublonmousse committed:
apply clippy lints

Doublonmousse committed:
remove drawing trashed and non trashed trees in visual debug

To circumvent gtk4 snapshot issues ...

Doublonmousse committed:
removed unneeded import and fmt
